File: DAYS/Day59-String encode and decode/encodeDecodeString.js
// strings - Encode and Decode Strings (leetcode 271) 
// Given a string, implement an algorithm to encode it and decode it.
// Example 1: Input: "Hello" Output: "5#Hello" 
// Example 2: Input: "Hello#World" Output: "11#Hello#World"
// Different Approaches:
// 1. Join delimiter - Time O(n), Space O(n)
// 2. Length delimiter - Time O(n), Space O(n)
// Leetcode :
// Solution :
 
function encode(strs) {
    return strs.map(s=>s.length+'#'+s).join('') // Join length and string with '#' delimiter
};

function decode(s){
    let res=[],i=0; // Initialize result array and index
    while(i<s.length){
        let j=i; // Find the position of the delimiter '#'
        while(s[j]!=='#')
            j++;
        let length=parseInt(s.slice(i,j));  // Extract the length
        res.push(s.slice(j+1,j+1+length)); // Extract the string segment
        i=j+1+length;  // Move to the next segment
    }
return res;
}
// Test Cases :
console.log(encode(["Hello","World"])); // "5#Hello5#World"
console.log(decode("5#Hello5#World")); // ["Hello","World"]
console.log(encode(["Hello#World"])); // "11#Hello#World"
console.log(decode("11#Hello#World")); // ["Hello#World"]

// Best Practices : Length delimiter 
// Complexity Analysis : Time Complexity: O(N) , Space Complexity: O(N)

File: DAYS/Day59-String encode and decode/encodeDecodeString.py
# String encoding and decoding leetcode(271)
# Given a string, design an algorithm to encode it and decode it.The encoded string is given as input to the decode function.
# Example 1: Input: "Hello" Output: "5#Hello" 
# Example 2: Input: "Hello World" Output: "5#Hello 5#World"
# Different Approach:
# 1. Join using Delimiter - Time O(N) Space O(N)
# 2. Length delimiter - Time O(N) Space O(N)
# Leetcode Link: https://leetcode.com/problems/encode-and-decode-strings/
# Solution :

def encode(strs):
    return ''.join(f'{len(s)}#{s}' for s in strs)  # '#'is add as a delimiter
def decode(s):
    res,i=[],0  # res to store result and i as index
    while i<len(s):
        j=i # j to find the delimiter '#'
        while s[j]!="#":
            j+=1 
        length=int(s[i:j]) # length of the string
        res.append(s[j+1:j+1+length]) # extract the string using length
        i=j+1+length # move i to the next string start
    return res
# Test cases
encoded_str = encode(["Hello","World", "I ","am"])
decoded_str = decode(encoded_str)
print(encoded_str) # "5#Hello5#World1#I2#am"
print(decoded_str) # ["Hello","World", "I ","am"]

# Best Approach: Length Delimiter
# Complexity Analysis: Time Complexity: O(N) , Space Complexity: O(N)
